src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeStackedItemContents.java

```
@@ -0,0 +1,129 @@
+        }
+    };
+
+    private final Object2ReferenceMap<ItemStack, ItemStack> stackInterner = new Object2ReferenceOpenCustomHashMap<>(STACK_HASH_STRATEGY);
+    private final StackedContents<ItemStack> stackedContents = new StackedContents<>();
+
+    private StackedContents.@Nullable Output<ItemStack> stackOutput;
+
+    @Override
+    public void accountStack(final ItemStack stack, final int maxStackSize) {
+        if (!stack.isEmpty()) {
+            // StackedContents works on Reference2IntMap, so if we meet stack which "same" copy
+            // has already been accounted we would need to provide the stack that was met first.
+            final ItemStack stackToAccount = this.stackInterner.computeIfAbsent(stack, Function.identity());
+            this.stackedContents.account(stackToAccount, Math.min(stack.getCount(), maxStackSize));
+        }
+    }
+
+    @Override
+    public boolean canCraft(
+        final Recipe<?> recipe, final int amount,
+        final StackedContents.@Nullable Output<Holder<Item>> output
+    ) {
+        final PlacementInfo placement = recipe.placementInfo();
+        return !placement.isImpossibleToPlace()
+            && this.stackedContents.tryPick(
+                ((PlacementInfoBridge) placement).bridge$getStackIngredientInfos(),
+                amount, this.createStackOutput(output));
+    }
+
+    @Override
+    public boolean canCraft(
+        final List<? extends StackedContents.IngredientInfo<Holder<Item>>> ingredients,
+        final StackedContents.@Nullable Output<Holder<Item>> output
+    ) {
+        // By default, this method is not called in the context the instance of this class is created.
+        // If this happens, it's either error in Sponge impl or
+        // mixin from some mod (which should be inspected instead of silently doing something that impl does not expect).
+        throw new UnsupportedOperationException("This method should not have been called, please report about it");
```

Lets just implement this

---

`List<? extends StackedContents.IngredientInfo<Holder<Item>>> ingredients` must be used on `StackedContents<Holder<Item>>`.

We only have `StackedContents<ItemStack>`, contents from `StackedItemContents` are not used currently, it may be changed and we can delegate to this (which in the context of this class doesn't really makes much sense but hides the exception).

In vanilla the only class that implements `StackedContents.IngredientInfo` is `Ingredient`, but I don't think it's safe to cast all of this to ingredients (because if this method is even called, it's done by some mod and such it can provide `IngredientInfo` other than `Ingredient`).

So, what way to implement this should be used?

---

I would expect us to just match the type because thats explicitly asked, we cant do anything else and the mod can't do anything more.